rslib/src/decks/limits.rs

```diff
@@ -71,40 +78,89 @@ impl RemainingLimits {
         config: &DeckConfig,
     ) -> RemainingLimits {
-        let (review_limit, new_limit) = if v3 {
-            let review_limit = normal
-                .current_review_limit(today)
-                .unwrap_or(config.inner.reviews_per_day);
-            let new_limit = normal
-                .current_new_limit(today)
-                .unwrap_or(config.inner.new_per_day);
-            (review_limit, new_limit)
-        } else {
-            (config.inner.reviews_per_day, config.inner.new_per_day)
-        };
-        let (new_today, mut rev_today) = deck.new_rev_counts(today);
         if v3 {
-            // any reviewed new cards contribute to the review limit
-            rev_today += new_today;
+            Self::new_for_normal_deck_v3(deck, today, normal, config)
+        } else {
+            Self::new_for_normal_deck_v2(deck, today, config)
         }
+    }
+
+    fn new_for_normal_deck_v2(deck: &Deck, today: u32, config: &DeckConfig) -> RemainingLimits {
+        let review_limit = config.inner.reviews_per_day;
+        let new_limit = config.inner.new_per_day;
+        let (new_today_count, review_today_count) = deck.new_rev_counts(today);
 
         Self {
-            review: (review_limit as i32 - rev_today).max(0) as u32,
-            new: (new_limit as i32 - new_today).max(0) as u32,
+            review: (review_limit as i32 - review_today_count).max(0) as u32,
+            new: (new_limit as i32 - new_today_count).max(0) as u32,
+            cap_new_to_review: false,
+        }
+    }
+
+    fn new_for_normal_deck_v3(
+        deck: &Deck,
+        today: u32,
+        normal: &NormalDeck,
+        config: &DeckConfig,
+    ) -> RemainingLimits {
+        let mut review_limit = normal
+            .current_review_limit(today)
+            .unwrap_or(config.inner.reviews_per_day) as i32;
+        let mut new_limit = normal
+            .current_new_limit(today)
+            .unwrap_or(config.inner.new_per_day) as i32;
+        let (new_today_count, review_today_count) = deck.new_rev_counts(today);
+
+        review_limit -= review_today_count;
+        new_limit -= new_today_count;
+        if !config.inner.new_ignore_review_limit {
+            review_limit -= new_today_count;
+            new_limit = new_limit.min(review_limit);
+        }
+
+        Self {
+            review: review_limit.max(0) as u32,
+            new: new_limit.max(0) as u32,
+            cap_new_to_review: !config.inner.new_ignore_review_limit,
+        }
+    }
+
+    pub(crate) fn get(&self, kind: LimitKind) -> u32 {
+        match kind {
+            LimitKind::Review => self.review,
+            LimitKind::New => self.new,
         }
     }
 
     pub(crate) fn cap_to(&mut self, limits: RemainingLimits) {
         self.review = self.review.min(limits.review);
         self.new = self.new.min(limits.new);
     }
+
+    /// True if some limit was decremented to 0.
+    fn decremented_to_zero(&mut self, kind: LimitKind) -> bool {
```

Comment:
I find this name a bit confusing. How about 'decrement' with a docstring that explains what the bool represents?

Reply:
I actually started out with `decrement`, but ain't happy with either name. I guess that means there's something wrong with the function!
Maybe I can return a struct instead and then call something like `limits.decrement(kind).decremented_to_zero`.